packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+    return ast::MK::UnresolvedConstant(core::LocOffsets::none(), move(scope), name);
+}
+
+ast::TreePtr PackageName::getFullNameLiteral(core::LocOffsets loc) const {
+    ast::TreePtr name = ast::MK::EmptyTree();
+    for (auto namePart : fullName) {
+        name = name2Expr(namePart, move(name));
+    }
+    // Outer name should have the provided loc.
+    if (auto lit = ast::cast_tree<ast::UnresolvedConstantLit>(name)) {
+        name = ast::MK::UnresolvedConstant(loc, move(lit->scope), lit->cnst);
+    }
+    return name;
+}
+
+ast::UnresolvedConstantLit *verifyConstant(core::MutableContext ctx, core::NameRef fun, ast::TreePtr &expr) {
+    auto target = ast::cast_tree<ast::UnresolvedConstantLit>(expr);
+    if (target == nullptr) {
+        if (auto e = ctx.beginError(expr->loc, core::errors::Packager::InvalidImportOrExport)) {
+            e.setHeader("Argument to `{}` must be a constant", fun.show(ctx));
+        }
+    }
+    return target;
+}
+
+core::LocOffsets firstLineOfFile(core::Context ctx) {
+    auto file = ctx.file;
+    core::LocOffsets firstLine{0, 0};
+    // Line 0 is reserved.
+    if (file.data(ctx).lineCount() > 1) {
+        firstLine.endLoc = static_cast<u4>(file.data(ctx).lineBreaks()[1]);
+    } else {
+        firstLine.endLoc = static_cast<u4>(file.data(ctx).source().length());
+    }
+    return firstLine;
+}
+
+struct PackageInfoFinder {
+    shared_ptr<PackageInfo> info = nullptr;
+    vector<unique_ptr<ast::UnresolvedConstantLit>> exported;
+    vector<unique_ptr<ast::UnresolvedConstantLit>> exportedMethods;
+    core::LocOffsets exportMethodsLoc = core::LocOffsets::none();
+
+    ast::TreePtr postTransformSend(core::MutableContext ctx, ast::TreePtr tree) {
+        auto *send = ast::cast_tree<ast::Send>(tree);
+
+        // Ignore methods
+        if (send->fun == core::Names::keepDef() || send->fun == core::Names::keepSelfDef()) {
+            return tree;
+        }
+
+        // Blacklisted methods
+        if (send->fun == core::Names::extend() || send->fun == core::Names::include()) {
+            if (auto e = ctx.beginError(send->loc, core::errors::Packager::InvalidPackageExpression)) {
+                e.setHeader("Invalid expression in package: `{}` is not allowed", send->fun.data(ctx)->shortName(ctx));
+            }
+            return tree;
+        }
+
+        // Sanity check arguments for unrecognized methods
+        if (send->fun != core::Names::export_() && send->fun != core::Names::import() &&
+            send->fun != core::Names::exportMethods()) {
+            for (const auto &arg : send->args) {
+                if (!ast::isa_tree<ast::Literal>(arg)) {
+                    if (auto e = ctx.beginError(arg->loc, core::errors::Packager::InvalidPackageExpression)) {
+                        e.setHeader("Invalid expression in package: Arguments to functions must be literals");
+                    }
+                }
+            }
+        }
```

Comment:
What methods could be called that aren't `export` / `import` / `exportMethods`?

Reply:
@nroman-stripe wants to be able to annotate packages with additional stuff, like `sorbet_min_typed_level 'true'`, which currently live in `package.yaml` files.

The plan is to have a `PackageSpec.rbi` within pay-server containing these extensions.